File: Frontend/validate_structure.js
/**
 * Frontend Structure Validation Script
 * Run with: node validate_structure.js
 */
const fs = require('fs');
const path = require('path');

const BASE_DIR = path.join(__dirname, 'src');

const REQUIRED_STRUCTURE = {
  'api': ['apiClient.ts', 'authApi.ts', 'petsApi.ts', 'chatApi.ts', 'adminApi.ts', 'index.ts'],
  'models': ['index.ts'],
  'lib': ['auth.tsx'],
  'components': [],
  'pages': ['auth', 'pets', 'admin'],
};

function checkDirectory(dirPath, requiredFiles = []) {
  const results = {
    exists: fs.existsSync(dirPath),
    files: [],
    missing: [],
  };

  if (!results.exists) {
    return results;
  }

  const items = fs.readdirSync(dirPath);
  
  if (requiredFiles.length > 0) {
    requiredFiles.forEach(file => {
      const filePath = path.join(dirPath, file);
      if (fs.existsSync(filePath)) {
        results.files.push(file);
      } else {
        results.missing.push(file);
      }
    });
  }

  return results;
}

function main() {
  console.log('='.repeat(60));
  console.log('Frontend Structure Validation');
  console.log('='.repeat(60));
  console.log();

  let allValid = true;

  // Check API layer
  console.log('Checking src/api/');
  const apiCheck = checkDirectory(path.join(BASE_DIR, 'api'), REQUIRED_STRUCTURE.api);
  if (apiCheck.exists && apiCheck.missing.length === 0) {
    console.log('  ✅ All API files present');
    apiCheck.files.forEach(file => console.log(`     ✓ ${file}`));
  } else {
    allValid = false;
    if (!apiCheck.exists) {
      console.log('  ❌ api/ directory does not exist');
    } else {
      console.log('  ❌ Missing files:');
      apiCheck.missing.forEach(file => console.log(`     ✗ ${file}`));
    }
  }
  console.log();

  // Check models
  console.log('Checking src/models/');
  const modelsCheck = checkDirectory(path.join(BASE_DIR, 'models'), REQUIRED_STRUCTURE.models);
  if (modelsCheck.exists && modelsCheck.missing.length === 0) {
    console.log('  ✅ Models present');
  } else {
    allValid = false;
    console.log('  ❌ Models missing');
  }
  console.log();

  // Check lib
  console.log('Checking src/lib/');
  const libCheck = checkDirectory(path.join(BASE_DIR, 'lib'), REQUIRED_STRUCTURE.lib);
  if (libCheck.exists && libCheck.missing.length === 0) {
    console.log('  ✅ Core libraries present');
  } else {
    allValid = false;
    console.log('  ❌ Core libraries missing');
  }
  console.log();

  // Check pages structure
  console.log('Checking src/pages/');
  const pagesPath = path.join(BASE_DIR, 'pages');
  if (fs.existsSync(pagesPath)) {
    REQUIRED_STRUCTURE.pages.forEach(pageDir => {
      const pagePath = path.join(pagesPath, pageDir);
      if (fs.existsSync(pagePath)) {
        console.log(`  ✅ pages/${pageDir}/ exists`);
      } else {
        allValid = false;
        console.log(`  ❌ pages/${pageDir}/ missing`);
      }
    });
  } else {
    allValid = false;
    console.log('  ❌ pages/ directory does not exist');
  }
  console.log();

  // Check package.json
  console.log('Checking project files:');
  const packageJson = path.join(__dirname, 'package.json');
  if (fs.existsSync(packageJson)) {
    console.log('  ✅ package.json');
    const pkg = JSON.parse(fs.readFileSync(packageJson, 'utf8'));
    if (pkg.dependencies && pkg.dependencies.axios) {
      console.log('  ✅ axios installed');
    } else {
      console.log('  ⚠️  axios not in dependencies');
    }
  } else {
    allValid = false;
    console.log('  ❌ package.json missing');
  }

  console.log();
  console.log('='.repeat(60));
  if (allValid) {
    console.log('✅ Structure is VALID - All files in place!');
  } else {
    console.log('❌ Structure has ISSUES - See above for details');
  }
  console.log('='.repeat(60));
}

main();

File: backend/validate_structure.py
#!/usr/bin/env python
"""
Structure validation script for Django backend.
Checks if all required files are in place.
"""
import os
from pathlib import Path

BASE_DIR = Path(__file__).parent

# Required files for each app
APP_REQUIRED_FILES = {
    'users': ['__init__.py', 'apps.py', 'models.py', 'views.py', 'serializers.py', 'urls.py', 'admin.py', 'backends.py'],
    'pets': ['__init__.py', 'apps.py', 'models.py', 'views.py', 'serializers.py', 'urls.py', 'admin.py'],
    'chats': ['__init__.py', 'apps.py', 'models.py', 'views.py', 'serializers.py', 'urls.py', 'admin.py'],
    'adminpanel': ['__init__.py', 'apps.py', 'models.py', 'views.py', 'serializers.py', 'urls.py', 'admin.py'],
    'api': ['__init__.py', 'apps.py', 'urls.py'],
}

# Required migrations
MIGRATIONS_REQUIRED = ['__init__.py']

def check_app_structure(app_name, required_files):
    """Check if app has all required files."""
    app_path = BASE_DIR / app_name
    missing = []
    existing = []
    
    if not app_path.exists():
        return False, [f"App folder '{app_name}' does not exist"], []
    
    for file in required_files:
        file_path = app_path / file
        if file_path.exists():
            existing.append(file)
        else:
            missing.append(file)
    
    return len(missing) == 0, missing, existing

def check_migrations(app_name):
    """Check if migrations folder exists and has __init__.py."""
    migrations_path = BASE_DIR / app_name / 'migrations'
    if not migrations_path.exists():
        return False, ["Migrations folder does not exist"]
    
    init_file = migrations_path / '__init__.py'
    if not init_file.exists():
        return False, ["migrations/__init__.py missing"]
    
    return True, []

def main():
    print("=" * 60)
    print("Django Backend Structure Validation")
    print("=" * 60)
    print()
    
    all_valid = True
    
    # Check each app
    for app_name, required_files in APP_REQUIRED_FILES.items():
        print(f"Checking {app_name}/")
        is_valid, missing, existing = check_app_structure(app_name, required_files)
        
        if is_valid:
            print(f"  ✅ All required files present")
            for file in existing:
                print(f"     ✓ {file}")
        else:
            all_valid = False
            print(f"  ❌ Missing files:")
            for file in missing:
                print(f"     ✗ {file}")
        
        # Check migrations
        if app_name != 'api':  # API doesn't need migrations
            migrations_valid, migrations_issues = check_migrations(app_name)
            if migrations_valid:
                print(f"  ✅ Migrations folder OK")
            else:
                all_valid = False
                print(f"  ❌ Migrations issues:")
                for issue in migrations_issues:
                    print(f"     ✗ {issue}")
        
        print()
    
    # Check project files
    print("Checking project files:")
    project_files = {
        'manage.py': BASE_DIR / 'manage.py',
        'requirements.txt': BASE_DIR / 'requirements.txt',
        'backend/settings.py': BASE_DIR / 'backend' / 'settings.py',
        'backend/urls.py': BASE_DIR / 'backend' / 'urls.py',
    }
    
    for name, path in project_files.items():
        if path.exists():
            print(f"  ✅ {name}")
        else:
            all_valid = False
            print(f"  ❌ {name} missing")
    
    print()
    print("=" * 60)
    if all_valid:
        print("✅ Structure is VALID - All files in place!")
    else:
        print("❌ Structure has ISSUES - See above for details")
    print("=" * 60)

if __name__ == '__main__':
    main()

